R/MultilabelDBRWrapper.R

```diff
@@ -0,0 +1,88 @@
+#' @title Use dependent binary relevance method (DBR) to create a multilabel learner.
+#'
+#' @description
+#' Every learner which is implemented in mlr and which supports binary
+#' classification can be converted to a wrapped DBR multilabel learner.
+#' The multilabel classification problem is converted into simple binary classifications
+#' for each label/target on which the binary learner is applied. 
+#' For each target, actual information of all binary labels (except the target variable) is used as additional features.
+#' During prediction these labels need are obtained by the binary relevance method using the same binary learner.
+#'
+#' Models can easily be accessed via \code{\link{getLearnerModel}}.
+#'
+#' @template arg_learner
+#' @template ret_learner
+#' @references
+#' Montanes, E. et al. (2013)
+#' \emph{Dependent binary relevance models for multi-label classification}
+#' Artificial Intelligence Center, University of Oviedo at Gijon, Spain.
+#' @family wrapper
+#' @family multilabel
+#' @export
+#' @examples
+#' d = getTaskData(yeast.task)
+#' # drop some labels so example runs faster
+#' d = d[, c(1:3, 15:117)]
+#' task = makeMultilabelTask(data = d, target = c("label1", "label2", "label3"))
+#' lrn = makeMultilabelDBRWrapper("classif.rpart")
+#' lrn = setPredictType(lrn, "prob")
+#' # train, predict and evaluate
+#' mod = train(lrn, task)
+#' pred = predict(mod, task)
+#' p = performance(pred)
+#' performance(pred, measure = hamloss)
+#' getMultilabelBinaryPerformances(pred, measures = list(mmce, auc))
+#' # above works also with predictions from resample!
+makeMultilabelDBRWrapper = function(learner) {
+  learner = checkLearner(learner, type = "classif")
+  hasLearnerProperties(learner, "twoclass")
+  id = paste("multilabel", learner$id, sep = ".")
+  packs = learner$package
+  x = makeHomogeneousEnsemble(id, learner$type, learner, packs,
+    learner.subclass = "MultilabelDBRWrapper",
+    model.subclass = "MultilabelDBRModel")
+  x$type = "multilabel"
+  return(x)
+}
+
+#' @export
+trainLearner.MultilabelDBRWrapper = function(.learner, .task, .subset, .weights = NULL, ...) {
+  targets = getTaskTargetNames(.task)
+  .task = subsetTask(.task, subset = .subset)
+  data = getTaskData(.task)
+  # train level 1 learners (binary relevance)
+  modelsLvl1 = getLearnerModel(train(makeMultilabelBinaryRelevanceWrapper(.learner$next.learner), .task, weights = .weights))
+  # train meta level learners
+  modelsMeta = namedList(targets)
+  for (tn in targets) {
+    data2Meta = dropNamed(data, setdiff(targets, tn))
+    trueLabelData = sapply(data.frame(data[, setdiff(targets, tn)]), as.numeric)
```

Comment:
Do you know this `data[, setdiff(targets, tn), drop=FALSE]`? It ensures
that it is always a data.frame, even if there is only one column. So please
use always drop=FALSE when you need data.frames for further computations.
Am 28.04.2016 18:52 schrieb "QuayAu" <notifications@github.com>:

> In R/MultilabelDBRWrapper.R
> <https://github.com/mlr-org/mlr/pull/833#discussion_r61462781>:
>
> > +  x$type = "multilabel"
> > +  return(x)
> > +}
> > +
> > +#' @export
> > +trainLearner.MultilabelDBRWrapper = function(.learner, .task, .subset, .weights = NULL, ...) {
> > +  targets = getTaskTargetNames(.task)
> > +  .task = subsetTask(.task, subset = .subset)
> > +  data = getTaskData(.task)
> > +  # train level 1 learners (binary relevance)
> > +  modelsLvl1 = getLearnerModel(train(makeMultilabelBinaryRelevanceWrapper(.learner$next.learner), .task, weights = .weights))
> > +  # train meta level learners
> > +  modelsMeta = namedList(targets)
> > +  for (tn in targets) {
> > +    data2Meta = dropNamed(data, setdiff(targets, tn))
> > +    trueLabelData = sapply(data.frame(data[, setdiff(targets, tn)]), as.numeric)
>
> data[, setdiff(targets, tn)] is only a data.frame, when there are more
> than 2 columns. In case of 1 column, it will be a logicalvector in this
> case (this happens, when there are only 2 labels).
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/mlr-org/mlr/pull/833/files/1c0cb9bbe645594daf28cbc2bc7e37e5b00cfb4b#r61462781>
>